pyodide_build/common.py

```diff
@@ -30,11 +32,48 @@
 
 
 def xbuildenv_dirname() -> str:
-    from pyodide_build import __version__
+    try:
+        from pyodide_build import __version__
+    except ImportError:
+        __version__ = "0.0.0"
 
     return f".pyodide-xbuildenv-{__version__}"
 
 
+@lru_cache(maxsize=1)
+def default_xbuildenv_path() -> Path:
+    """
+    Return the default path to the cross-build environment directory.
+
+    This directory is used when no path is provided to the `pyodide xbuildenv` subcommands.
+    """
+    dirname = xbuildenv_dirname()
+    candidates = []
+
+    # 1. default cache directory
+    candidates.append(Path(platformdirs.user_cache_dir()) / dirname)
+
+    # 2. current working directory
+    candidates.append(Path.cwd() / dirname)
+
+    for candidate in candidates:
+        if _has_write_access(candidate):
+            return candidate
+
+
+def _has_write_access(folder: Path) -> bool:
+    """
+    Checks if the current user has write access to the given folder using pathlib.
+    """
+    try:
+        if not folder.exists() and folder.parent != folder:
+            return _has_write_access(folder.parent)
+
+        return os.access(str(folder), os.W_OK)
```

Comment:
Should we check for `W_OK | X_OK`? As far as I understand, we can't add a file to a folder without execute permissions?

Reply:
> Should we check for `W_OK | X_OK`? As far as I understand, we can't add a file to a folder without execute permissions?

I think `W_OK` is fine. What we should check for are (assuming `Path` is the absolute/resolved path to the directory):

- if `Path.owner() == os.getlogin()` returns `True` (checks for the owner by name); and
- if `Path.stat().st_uid == os.geteuid()` returns `True` (checks for the effective user ID).

These might not be available everywhere (https://docs.python.org/3/library/os.html#os.geteuid), but no one will run `pyodide-build` in WASI anyway.